File: bb-pm-tools/src/rate-limit.ts
import { config } from "./config";
import { getRedis } from "./redis";

/**
 * Token-bucket rate limit for /agent/run. Uses Redis if available so the
 * limit is shared across plugin instances; otherwise falls back to a
 * per-process in-memory bucket. The cap stays small (~30 req/min by
 * default) because each /agent/run hits the LLM and bb-pm DB.
 *
 * Limit is keyed by best-effort client identifier: prefer the channel-
 * supplied `correlationId` if present (one identifier per chat thread),
 * else the request IP. This protects against runaway loops more than
 * malicious abuse — a real DoS layer belongs upstream of the gateway.
 *
 * Redis encoding: `${windowStart}:${count}` so we can compute remaining
 * window time without a second PTTL call. Keeps the wire format simple.
 */

const KEY_PREFIX = "bbpm:rl:agent-run:";

type Bucket = { count: number; windowStart: number };
const memBuckets = new Map<string, Bucket>();

export type RateLimitResult = {
  ok: boolean;
  remaining: number;
  retryAfterSec: number;
  limit: number;
};

export async function checkRateLimit(identifier: string): Promise<RateLimitResult> {
  const limit = config.rateLimit.maxPerWindow;
  if (limit <= 0) {
    return { ok: true, remaining: Number.POSITIVE_INFINITY, retryAfterSec: 0, limit };
  }
  const winSec = config.rateLimit.windowSec;
  const winMs = winSec * 1000;
  const now = Date.now();

  const r = config.redis.url ? await getRedis() : null;
  if (r) {
    const key = KEY_PREFIX + identifier;
    try {
      const raw = await r.get(key);
      let windowStart = now;
      let count = 0;
      if (raw) {
        const [wsStr, cStr] = raw.split(":");
        const ws = Number(wsStr);
        const c = Number(cStr);
        if (Number.isFinite(ws) && Number.isFinite(c) && now - ws < winMs) {
          windowStart = ws;
          count = c;
        }
      }
      if (count >= limit) {
        const elapsed = now - windowStart;
        return {
          ok: false,
          remaining: 0,
          retryAfterSec: Math.max(1, Math.ceil((winMs - elapsed) / 1000)),
          limit,
        };
      }
      const ttlSec = Math.max(1, Math.ceil((winMs - (now - windowStart)) / 1000));
      await r.set(key, `${windowStart}:${count + 1}`, "EX", ttlSec).catch(() => {});
      return { ok: true, remaining: limit - count - 1, retryAfterSec: 0, limit };
    } catch {
      // fall through to in-memory
    }
  }

  const b = memBuckets.get(identifier);
  if (!b || now - b.windowStart > winMs) {
    memBuckets.set(identifier, { count: 1, windowStart: now });
    return { ok: true, remaining: limit - 1, retryAfterSec: 0, limit };
  }
  if (b.count >= limit) {
    const elapsed = now - b.windowStart;
    return {
      ok: false,
      remaining: 0,
      retryAfterSec: Math.max(1, Math.ceil((winMs - elapsed) / 1000)),
      limit,
    };
  }
  b.count += 1;
  return { ok: true, remaining: limit - b.count, retryAfterSec: 0, limit };
}

File: agent/orchestrator/rate_limit.py
"""Per-caller rate limiting. Ports bb-pm-tools/src/rate-limit.ts.

Fixed-window counter, Redis-backed when available, in-process otherwise.
Key is `cid:<correlationId>` (preferred) or `ip:<addr>`.
"""

from __future__ import annotations

import time
from dataclasses import dataclass

from core.config import settings
from infrastructure.redis import redis_client


@dataclass
class RateLimitResult:
    ok: bool
    limit: int
    remaining: int
    retry_after_sec: int


_memory: dict[str, tuple[int, int]] = {}  # key -> (window_start, count)


async def check_rate_limit(key: str) -> RateLimitResult:
    limit = settings.rate_limit.max_per_window
    window = settings.rate_limit.window_sec
    if limit <= 0:
        return RateLimitResult(ok=True, limit=limit, remaining=limit, retry_after_sec=0)

    now = int(time.time())
    window_index = now // window
    redis = redis_client.client

    if redis is not None:
        redis_key = f"rl:{key}:{window_index}"
        try:
            count = await redis.incr(redis_key)
            if count == 1:
                await redis.expire(redis_key, window)
        except Exception:  # noqa: BLE001 — fall back to in-memory
            count = _memory_incr(key, window_index)
    else:
        count = _memory_incr(key, window_index)

    remaining = max(0, limit - count)
    if count > limit:
        retry_after = window - (now % window)
        return RateLimitResult(ok=False, limit=limit, remaining=0, retry_after_sec=retry_after)
    return RateLimitResult(ok=True, limit=limit, remaining=remaining, retry_after_sec=0)


def _memory_incr(key: str, window_index: int) -> int:
    start, count = _memory.get(key, (window_index, 0))
    if start != window_index:
        start, count = window_index, 0
    count += 1
    _memory[key] = (start, count)
    if len(_memory) > 5000:  # lazy GC
        for k, (s, _) in list(_memory.items()):
            if s != window_index:
                _memory.pop(k, None)
    return count
